Fix: compatibility with BLE and NimBLE in today's current versions

JuniorPolegato opened the pull request:
Hi!

I made some adjustments for compatibility with new versions of BLE and NimBLE.

JuniorPolegato committed:
Fix: compatibility with BLE and NimBLE in today's current versions

JuniorPolegato committed:
Fix: NimBLE advertiseOnDisconnect and Security

JuniorPolegato committed:
Fix: security and authentication issues on some android devices

awong1900 commented:
That is good job. Do this branch support nimble v2.3.x and esp arduino 3.0?

JuniorPolegato commented:
> That is good job. Do this branch support nimble v2.3.x and esp arduino 3.0?

Hi! Thanks.

Yes, I compiled with "esp32 by Espressif Systems Version 3.2.1" and with (and without) "NimBLE-Arduino by H2zero Version 2.3.2" and  this test code:
```
#include <Arduino.h>
#include <BleKeyboard.h>

BleKeyboard bleKeyboard;

void setup() {
  Serial.begin(115200, SERIAL_8N1);
  while (!Serial);

  Serial.println("\n\n\n\nBLE Keyborad test...\n\n\n\n\n");

  bleKeyboard.setName("Keyboard");
  bleKeyboard.begin();
  bleKeyboard.setDelay(30);
}

void loop() {
  static size_t buffer_len;
  if (bleKeyboard.isConnected()) {
    if ((buffer_len = Serial.available())) {
      char data[buffer_len + 1];
      Serial.read(data, buffer_len);
      data[buffer_len] = '\0';
      Serial.print(data);
      Serial.flush();
      bleKeyboard.print(data);
      bleKeyboard.flush();
    }
    else {
      Serial.print("-");
    }
  }
  else {
    Serial.print(".");
  }
  Serial.flush();
  delay(300);
}
```

StuDentBRz commented:
Sorry for being noob where can I download the updated version of the the library that you made? Tried to download on your github page by clicking on code and download zip, installed the library on ide but got an error while was compiling the code on it...

JuniorPolegato commented:
Hi!

You need to extract the zip file in the libraries directory for your user, in my case it is "/home/junior/projects/arduino/libraries".
Search on your settings.
I use VSCodium with Arduino Community extension, instead of ArduinoIDE or PlatformIO.

JuniorPolegato commented:
Hi!

I updated to "esp32 by Espressif Systems Version 3.3.0", this fix was necessary:
https://github.com/espressif/arduino-esp32/pull/11681
I manually modified the file `~/.arduino15/packages/esp32/hardware/esp32/3.3.0/libraries/BLE/src/BLESecurity.cpp`, adding the new lines from this fix.

I also updated to "NimBLE-Arduino by H2zero Version 2.3.4".

I can now confirm that the code works fine in these versions with "#define USE_NIMBLE" or "#undef USE_NIMBLE".

McCarthy83 commented:
> Hi!
> 
> I updated to "esp32 by Espressif Systems Version 3.3.0", this fix was necessary: [espressif/arduino-esp32#11681](https://github.com/espressif/arduino-esp32/pull/11681) I manually modified the file `~/.arduino15/packages/esp32/hardware/esp32/3.3.0/libraries/BLE/src/BLESecurity.cpp`, adding the new lines from this fix.
> 
> I also updated to "NimBLE-Arduino by H2zero Version 2.3.4".
> 
> I can now confirm that the code works fine in these versions with "#define USE_NIMBLE" or "#undef USE_NIMBLE".

Following your commit changing the code (thank you by the way) I try to compile your code and get this error:

ESP32_BLE_Keyboard/BleKeyboard.cpp:158:14: error: 'class BLESecurity' has no member named 'setStaticPIN'
  158 |   pSecurity->setStaticPIN(112233);  // <--- It has to be set before the lines below as suggested by the setStaticPIN() definition comments
      |              ^~~~~~~~~~~~
exit status 1

Compilation error: exit status 1

JuniorPolegato commented:
> > Hi!
> > I updated to "esp32 by Espressif Systems Version 3.3.0", this fix was necessary: [espressif/arduino-esp32#11681](https://github.com/espressif/arduino-esp32/pull/11681) I manually modified the file `~/.arduino15/packages/esp32/hardware/esp32/3.3.0/libraries/BLE/src/BLESecurity.cpp`, adding the new lines from this fix.
> > I also updated to "NimBLE-Arduino by H2zero Version 2.3.4".
> > I can now confirm that the code works fine in these versions with "#define USE_NIMBLE" or "#undef USE_NIMBLE".
> 
> Following your commit changing the code (thank you by the way) I try to compile your code and get this error:
> 
> ESP32_BLE_Keyboard/BleKeyboard.cpp:158:14: error: 'class BLESecurity' has no member named 'setStaticPIN' 158 | pSecurity->setStaticPIN(112233); // <--- It has to be set before the lines below as suggested by the setStaticPIN() definition comments | ^~~~~~~~~~~~ exit status 1
> 
> Compilation error: exit status 1

Hi!
What version of esp32 are you using?
I compiled with 3.3.0, and the patch applied, and finished well.

JuniorPolegato committed:
Fix: for version 3.3.2, replaced setStaticPIN(<key>) by setPassKey(false)

JuniorPolegato commented:
> > > Hi!
> > > I updated to "esp32 by Espressif Systems Version 3.3.0", this fix was necessary: [espressif/arduino-esp32#11681](https://github.com/espressif/arduino-esp32/pull/11681) I manually modified the file `~/.arduino15/packages/esp32/hardware/esp32/3.3.0/libraries/BLE/src/BLESecurity.cpp`, adding the new lines from this fix.
> > > I also updated to "NimBLE-Arduino by H2zero Version 2.3.4".
> > > I can now confirm that the code works fine in these versions with "#define USE_NIMBLE" or "#undef USE_NIMBLE".
> > 
> > 
> > Following your commit changing the code (thank you by the way) I try to compile your code and get this error:
> > ESP32_BLE_Keyboard/BleKeyboard.cpp:158:14: error: 'class BLESecurity' has no member named 'setStaticPIN' 158 | pSecurity->setStaticPIN(112233); // <--- It has to be set before the lines below as suggested by the setStaticPIN() definition comments | ^~~~~~~~~~~~ exit status 1
> > Compilation error: exit status 1
> 
> Hi! What version of esp32 are you using? I compiled with 3.3.0, and the patch applied, and finished well.

Hi!
For `arduino-esp32` version `3.3.2`, I found that `setStaticPIN(<key>)` has been replaced with `setPassKey(true, <key>)`, or for a random key using `setPassKey(false)`.
So I updated the code with the second one.
Please, pull the new commit and try again.

McCarthy83 commented:
Awesome, so I first apply the commit for 3.3.0 and then change the most recent one? Thank you so much by the way.

JuniorPolegato commented:
Hi!
Just update ESP32 board to 3.3.2 and get the last commit/version from https://github.com/JuniorPolegato/ESP32-BLE-Keyboard
